Datatype selection option for roofline

cfallows-amd opened the pull request:
Added command line option to specify which datatype(s) to capture into the roofline PDF(s)
All datatypes are still collected by roofline call if applicable, but only specific datatypes are plotted into PDF outputs
Default is FP32

Reworked roofline calls and plotting to be general enough such that any new datatypes added into rocm-amdgpu-bench can easily be reflected in rocprof-compute with simple modifications in roofline_calc.py

Adjusted ctest to reflect expected default pdf outputs from roofline

cfallows-amd committed:
Reformatting checks and execution of peak and mfma calculations for supported archs and datatypes

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
Add datatype selection option to argparser
- able to take as many datatypes as long as they are valid, gives error otherwise
- will skip datatype and give error message if the datatype is not valid on a particular gpu arch
- will dump out all datatypes into one graph, but separate FP from Int into two graphs if needed
- easier to add new datatypes now by just adding available datatypes to the lists of supported DT per arch and if it has peak and/or mfma values

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
Adjust ctests ROOF_ONLY_FILES to reflect default roofline output changes to only expect F32 plot pdf

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
Add descritpion to table to notify users where table contents come from for roofline datatype support

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
Remove colour hardcoding for AI points
Edit arg description for datatype selection

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
Merge branch 'develop' into cfallows/select_datatype

cfallows-amd commented:
Ran ctest and manual testing on MI210 and MI300X systems.

feizheng10 reviewed (approved):
LG. Changelog please.
And I guess there will some doc work to do:)

feizheng10 commented on src/argparser.py:
```diff
@@ -367,6 +367,19 @@ def validate_block(value):
+
+    roofline_group.add_argument(
+        "-D",
+        "--datatype",
```
I'd suggest to use "--roofline-data-type". In the future, we need this for post-analysis as well.

cfallows-amd replied:
Agree, will make change.

feizheng10 commented:
User still could do combo of multiple data types in one plot, right?

cfallows-amd commented:
Yes, combo plot occurs if more than one of a type (FP vs Int) gets selected, there will be more than one type on the graph. Int and FP graphs will always be separated because of AI for FP. Docs will highlight that one run of roofline will collect all data, so there is still an option to generate separate plots for each type off of the same data collection run if there exists a workload directory already populated. I will work with docs team to update the roofline section today. Thanks

cfallows-amd committed:
Modify option naming
Update CHANGELOG
Update docs for roofline datatype selection usage

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
Fix arg option processing
Fix no exit on error message for invalid datatypes on arch
(Manual testing on mi200/mi300)

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
Merge branch 'develop' into cfallows/select_datatype
